### PR TITLE
add uTP support through rust-utp

### PR DESCRIPTION
**Pull request by vinipsmaker**
TODO:

- Use upstream rust-utp.

[Review on Reviewable](https://reviewable.io/reviews/maidsafe/crust/238)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @dirvine

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/utp_connections.rs [hunk 1]
```diff
@@ -31,24 +35,77 @@ pub type OutUtpStream<T> = Sender<T>;
...
-    panic!("UTP support in crust is currently non functional");
+    let socket = UtpWrapper::wrap(newconnection);
+    let output = socket.output();
+    Ok((upgrade_reader(socket), upgrade_writer(output)))
 }
 
+fn upgrade_writer<'a, T>(sender: Sender<Vec<u8>>) -> Sender<T>
+    where T: Send + Encodable + 'static {
+    let (tx, rx) = mpsc::channel();
+    let _ = thread::spawn(move || {
```
Comment (chandraprakash) [inline src/utp_connections.rs]: May be worth using `thread::Builder`. Helps in debugging.

**Comment by vinipsmaker** [comment]
Current list of issues:

- `Encodable`/`Decodable` don't care about uTP.
- It only listens on TCP by default.
- Update code to better support hole punching.

**Comment by vinipsmaker** [comment]
New current list of issues:

- `Encodable`/`Decodable` don't care about uTP.
- Update code to better support hole punching (can be done after merging if needed).

**Comment by chandraprakash** [comment]
Encodable/Decodable don't care about uTP.
This is already done by Niall in uTP integration.
See this issue https://github.com/maidsafe/crust/issues/249

**Comment by vinipsmaker** [comment]
It should be okay to merge this PR now.

**Pull request merged**